File: rs/tests/test_move.rs
mod common;

use gitstore::*;

fn store_with_move_files(dir: &std::path::Path) -> (GitStore, Fs) {
    let store = common::create_store(dir, "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write("hello.txt", b"hello world").unwrap();
    batch.write("dir/a.txt", b"aaa").unwrap();
    batch.write("dir/b.txt", b"bbb").unwrap();
    batch.write("other/c.txt", b"ccc").unwrap();
    batch.commit().unwrap();
    let fs = store.branches().get("main").unwrap();
    (store, fs)
}

// ---------------------------------------------------------------------------
// Rename file
// ---------------------------------------------------------------------------

#[test]
fn rename_file_basic() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("hello.txt", "renamed.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("renamed.txt").unwrap(), "hello world");
    assert!(!fs.exists("hello.txt").unwrap());
}

#[test]
fn rename_preserves_other_files() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("hello.txt", "renamed.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("dir/a.txt").unwrap(), "aaa");
    assert_eq!(fs.read_text("dir/b.txt").unwrap(), "bbb");
    assert_eq!(fs.read_text("other/c.txt").unwrap(), "ccc");
}

// ---------------------------------------------------------------------------
// Rename directory
// ---------------------------------------------------------------------------

#[test]
fn rename_directory_basic() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("dir", "newdir", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert_eq!(fs.read_text("newdir/a.txt").unwrap(), "aaa");
    assert_eq!(fs.read_text("newdir/b.txt").unwrap(), "bbb");
    assert!(!fs.exists("dir/a.txt").unwrap());
    assert!(!fs.exists("dir/b.txt").unwrap());
}

#[test]
fn rename_directory_preserves_others() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("dir", "moved", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    // hello.txt and other/ should still be there
    assert_eq!(fs.read_text("hello.txt").unwrap(), "hello world");
    assert_eq!(fs.read_text("other/c.txt").unwrap(), "ccc");
}

// ---------------------------------------------------------------------------
// Rename to different directory
// ---------------------------------------------------------------------------

#[test]
fn rename_file_to_new_directory() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("hello.txt", "newdir/hello.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("hello.txt").unwrap());
    assert_eq!(fs.read_text("newdir/hello.txt").unwrap(), "hello world");
}

#[test]
fn rename_nested_file() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("dir/a.txt", "dir/renamed_a.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("dir/a.txt").unwrap());
    assert_eq!(fs.read_text("dir/renamed_a.txt").unwrap(), "aaa");
    // b.txt still there
    assert_eq!(fs.read_text("dir/b.txt").unwrap(), "bbb");
}

#[test]
fn rename_file_across_directories() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("dir/a.txt", "other/moved_a.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("dir/a.txt").unwrap());
    assert_eq!(fs.read_text("other/moved_a.txt").unwrap(), "aaa");
    assert_eq!(fs.read_text("other/c.txt").unwrap(), "ccc");
}

// ---------------------------------------------------------------------------
// Atomicity
// ---------------------------------------------------------------------------

#[test]
fn rename_is_single_commit() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    let log_before = fs.log(Default::default()).unwrap();
    fs.rename("hello.txt", "moved.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    let log_after = fs.log(Default::default()).unwrap();
    assert_eq!(log_after.len(), log_before.len() + 1);
}

#[test]
fn rename_previous_commit_has_original() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("hello.txt", "moved.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    let prev = fs.back(1).unwrap();
    assert!(prev.exists("hello.txt").unwrap());
    assert!(!prev.exists("moved.txt").unwrap());
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

#[test]
fn rename_nonexistent_source_errors() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = store_with_move_files(dir.path());
    assert!(fs.rename("missing.txt", "dest.txt", Default::default()).is_err());
}

// ---------------------------------------------------------------------------
// Custom message
// ---------------------------------------------------------------------------

#[test]
fn rename_custom_message() {
    let dir = tempfile::tempdir().unwrap();
    let (store, fs) = store_with_move_files(dir.path());
    fs.rename("hello.txt", "moved.txt", fs::WriteOptions {
        message: Some("renamed hello".into()),
        ..Default::default()
    })
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    let log = fs.log(fs::LogOptions { limit: Some(1), ..Default::default() }).unwrap();
    assert_eq!(log[0].message, "renamed hello");
}

// ---------------------------------------------------------------------------
// Stale
// ---------------------------------------------------------------------------

#[test]
fn rename_on_stale_errors() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("a.txt", b"a", Default::default()).unwrap();
    let fs_old = store.branches().get("main").unwrap();

    // Advance
    let fs_new = store.branches().get("main").unwrap();
    fs_new.write("b.txt", b"b", Default::default()).unwrap();

    assert!(fs_old.rename("a.txt", "moved.txt", Default::default()).is_err());
}

// ---------------------------------------------------------------------------
// Directory rename preserves content
// ---------------------------------------------------------------------------

#[test]
fn rename_directory_deep() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write("src/a/b/c.txt", b"deep").unwrap();
    batch.write("src/x.txt", b"x").unwrap();
    batch.commit().unwrap();

    let fs = store.branches().get("main").unwrap();
    fs.rename("src", "dest", Default::default()).unwrap();

    let fs = store.branches().get("main").unwrap();
    assert!(!fs.exists("src").unwrap());
    assert_eq!(fs.read_text("dest/a/b/c.txt").unwrap(), "deep");
    assert_eq!(fs.read_text("dest/x.txt").unwrap(), "x");
}

File: cpp/tests/test_move.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_mtest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

// ---------------------------------------------------------------------------
// move: basic rename
// ---------------------------------------------------------------------------

TEST_CASE("move: simple rename", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("old.txt", "data");

    snap = snap.move({"old.txt"}, "new.txt");
    CHECK(snap.read_text("new.txt") == "data");
    CHECK_FALSE(snap.exists("old.txt"));
    fs::remove_all(path);
}

TEST_CASE("move: into directory", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");
    snap = snap.write_text("dir/existing.txt", "existing");

    snap = snap.move({"file.txt"}, "dir");
    CHECK(snap.read_text("dir/file.txt") == "data");
    CHECK(snap.read_text("dir/existing.txt") == "existing");
    CHECK_FALSE(snap.exists("file.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: multiple sources
// ---------------------------------------------------------------------------

TEST_CASE("move: multiple files into directory", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    auto batch = snap.batch();
    batch.write_text("a.txt", "aaa");
    batch.write_text("b.txt", "bbb");
    batch.write_text("dest/placeholder.txt", "p");
    snap = batch.commit();

    snap = snap.move({"a.txt", "b.txt"}, "dest");
    CHECK(snap.read_text("dest/a.txt") == "aaa");
    CHECK(snap.read_text("dest/b.txt") == "bbb");
    CHECK_FALSE(snap.exists("a.txt"));
    CHECK_FALSE(snap.exists("b.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: directory
// ---------------------------------------------------------------------------

TEST_CASE("move: rename file to nested path", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    snap = snap.move({"file.txt"}, "sub/renamed.txt");
    CHECK(snap.read_text("sub/renamed.txt") == "data");
    CHECK_FALSE(snap.exists("file.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: dry run
// ---------------------------------------------------------------------------

TEST_CASE("move: dry run does not modify original", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    vost::MoveOptions mopts;
    mopts.dry_run = true;
    snap.move({"file.txt"}, "renamed.txt", mopts);

    // dry_run: original file still exists
    CHECK(snap.exists("file.txt"));
    // Re-read from store to confirm no commit happened
    auto snap2 = store.branches().get("main");
    CHECK(snap2.exists("file.txt"));
    CHECK_FALSE(snap2.exists("renamed.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: read-only rejection
// ---------------------------------------------------------------------------

TEST_CASE("move: read-only Fs throws", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    store.tags().set("v1", snap);
    auto tag_snap = store.tags().get("v1");

    REQUIRE_THROWS_AS(tag_snap.move({"file.txt"}, "new.txt"), vost::PermissionError);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: preserves other files
// ---------------------------------------------------------------------------

TEST_CASE("move: preserves other files", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "aaa");
    snap = snap.write_text("other.txt", "other");

    snap = snap.move({"a.txt"}, "b.txt");
    CHECK_FALSE(snap.exists("a.txt"));
    CHECK(snap.read_text("b.txt") == "aaa");
    CHECK(snap.read_text("other.txt") == "other");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: directory recursive
// ---------------------------------------------------------------------------

TEST_CASE("move: directory recursive", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("src/a.txt", "a");
    snap = snap.write_text("src/sub/b.txt", "b");

    vost::MoveOptions opts;
    opts.recursive = true;
    snap = snap.move({"src"}, "dst", opts);
    CHECK_FALSE(snap.exists("src"));
    CHECK(snap.read_text("dst/a.txt") == "a");
    CHECK(snap.read_text("dst/sub/b.txt") == "b");
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: dry_run reports correct paths
// ---------------------------------------------------------------------------

TEST_CASE("move: dry_run reports correct paths", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "data");
    snap = snap.write_text("b.txt", "other");

    vost::MoveOptions mopts;
    mopts.dry_run = true;
    auto result = snap.move({"a.txt"}, "renamed.txt", mopts);

    // Original unchanged
    CHECK(snap.exists("a.txt"));
    // No commit happened
    auto snap2 = store.branches().get("main");
    CHECK(snap2.exists("a.txt"));
    CHECK_FALSE(snap2.exists("renamed.txt"));
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: error cases
// ---------------------------------------------------------------------------

TEST_CASE("move: nonexistent source throws", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("exists.txt", "data");

    REQUIRE_THROWS_AS(snap.move({"ghost.txt"}, "dest.txt"), vost::NotFoundError);
    fs::remove_all(path);
}

TEST_CASE("move: directory without recursive throws", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("dir/file.txt", "content");

    REQUIRE_THROWS_AS(snap.move({"dir"}, "other"), vost::IsADirectoryError);
    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// move: custom message
// ---------------------------------------------------------------------------

TEST_CASE("move: custom commit message", "[move]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "data");

    vost::MoveOptions mopts;
    mopts.message = "custom move message";
    snap = snap.move({"a.txt"}, "b.txt", mopts);

    CHECK(snap.message() == "custom move message");
    fs::remove_all(path);
}
